### PR TITLE
Additional cleanup for Python extension

### PR DESCRIPTION
**Pull request by pepone**
Includes #2453 fixes

### CODE REVIEW

File: cpp/src/slice2py/PythonUtil.cpp [hunk 1]
```diff
@@ -1776,14 +1763,8 @@ Slice::Python::CodeVisitor::writeAssign(const MemberInfo& info)
...
-        _out << nl << "if " << paramName << " is Ice._struct_marker:";
-        _out.inc();
-        _out << nl << "self." << memberName << " = " << getSymbol(st) << "()";
-        _out.dec();
-        _out << nl << "else:";
-        _out.inc();
-        _out << nl << "self." << memberName << " = " << paramName;
-        _out.dec();
+        _out << nl << "self." << memberName << " = " << paramName << "if " << paramName << " is not None else "
+             << getSymbol(st) << "()";
```
Comment (pepone) [inline cpp/src/slice2py/PythonUtil.cpp:1767]: I replaced `_struct_marker` with `None` it is not clear why we need to have default values for all parameters. I need to review the generated code for constructors.

File: python/modules/IcePy/Init.cpp [hunk 2]
```diff
@@ -171,7 +165,7 @@ static struct PyModuleDef iceModule = {
     nullptr,
     nullptr,
     nullptr,
-    nullptr};
+    IcePy_cleanup};
```
Comment (pepone) [inline python/modules/IcePy/Init.cpp:168]: We were calling the cleanup function from Ice/__init__.py. I move this to the module-free function.

File: python/modules/IcePy/Logger.cpp [hunk 3]
```diff
@@ -414,7 +414,7 @@ IcePy::cleanupLogger()
     //
     // Python is about to exit; we need to remove the wrapper around the process logger.
     //
-    Ice::setProcessLogger(0);
+    Ice::setProcessLogger(nullptr);
```
Comment (pepone) [inline python/modules/IcePy/Logger.cpp:417]: That is the only thing done by the cleanup, not clear we need it at all.

File: python/modules/IcePy/ObjectAdapter.cpp [hunk 4]
```diff
@@ -1807,7 +1807,7 @@ IcePy::wrapObjectAdapter(const Ice::ObjectAdapterPtr& adapter)
     {
         return 0;
     }
-    PyObject* wrapperType = lookupType("Ice.ObjectAdapterI");
+    PyObject* wrapperType = lookupType("Ice.ObjectAdapterI.ObjectAdapterI");
```
Comment (pepone) [inline python/modules/IcePy/ObjectAdapter.cpp:1810]: I removed the import ObjectAdaterI from Ice/__init__.py as the file is not part of the public API, is just a wrapper used by IcePy.

File: python/python/Ice/Current.py
Comment (pepone) [inline python/python/Ice/Current.py:1]: This was not used, the actual definition for Current is in IcePy C++ code.

File: python/python/Ice/Object.py [hunk 5]
```diff
@@ -95,3 +98,51 @@ def handler(future):
             cb.response(ex.value)
         except BaseException:
             cb.exception(sys.exc_info()[1])
+
+IcePy._t_Object = IcePy.defineClass("::Ice::Object", Object, (), None, ())
```
Comment (pepone) [inline python/python/Ice/Object.py:102]: Move here from Ice/__init__.py

File: python/python/Ice/UnknownSlicedValue.py [hunk 6]
```diff
@@ -33,3 +34,8 @@ class SliceInfo(object):
     # hasOptionalMembers - boolean
     # isLastSlice - boolean
     pass
+
+IcePy._t_UnknownSlicedValue = IcePy.defineValue(
```
Comment (pepone) [inline python/python/Ice/UnknownSlicedValue.py:38]: moved here from Ice/__init__.py

File: python/python/Ice/Value.py [hunk 7]
```diff
@@ -30,3 +32,5 @@ def ice_getSlicedData(self):
         Returns:
             The sliced data or null."""
         return getattr(self, "_ice_slicedData", None)
+
+IcePy._t_Value = IcePy.defineValue("::Ice::Object", Value, -1, (), False, None, ())
```
Comment (pepone) [inline python/python/Ice/Value.py:36]: moved here from Ice/__init__.py

File: python/python/Ice/__init__.py [hunk 8]
```diff
@@ -1,21 +1,14 @@
...
 """
 
-import array
-import atexit
 import IcePy
-
-#
-# Give the extension an opportunity to clean up before a graceful exit.
-#
-atexit.register(IcePy.cleanup)
```
Comment (pepone) [inline python/python/Ice/__init__.py:18]: The cleanup was moved to Init.cpp

File: python/python/Ice/__init__.py [hunk 9]
```diff
@@ -35,214 +28,97 @@
...
-from .ModuleUtil import *
-
 #
-# This value is used as the default value for struct types in the constructors
-# of user-defined types. It allows us to determine whether the application has
-# supplied a value. (See bug 3676)
+# Forward declarations.
 #
-# TODO: can we use None instead?
-_struct_marker = object()
```
Comment (pepone) [inline python/python/Ice/__init__.py:54]: Removed, replaced by None

See the mentioned bug https://zeroc.atlassian.net/jira/software/c/projects/ICE/issues/ICE-3381?jql=project%20%3D%20%22ICE%22%20AND%20%22bug%20import%20id%5Bshort%20text%5D%22%20~%20%223676%22%20ORDER%20BY%20created%20DESC

**Review by bernardnormier — APPROVED** [review]

**Review by externl — APPROVED** [review]

File: python/modules/IcePy/Logger.cpp (hunk 3 above)
Comment (pepone) [reply python/modules/IcePy/Logger.cpp]: I don't think so. At this point we are unloading the extension nothing should be using the process logger.

**Pull request merged**